contracts/modules/Checkpoint/Voting/APLCR/AdvancedPLCRVotingCheckpoint.sol

```diff
@@ -0,0 +1,1003 @@
+        ) {
+            return (ballot.totalProposals * securityToken.balanceOfAt(_voter, ballot.checkpointId));
+        } else {
+            // Return voting power = 0, when voter already vote for the given ballotId
+            return uint256(0);
+        }
+    }
+
+    /**
+     * @notice Queries the result of a given ballot
+     * @dev should be called off-chain
+     * @param _ballotId Id of the target ballot
+     * @return uint256 choicesWeighting
+     * @return uint256 noOfChoicesInProposal
+     * @return address voters
+     */
+    function getBallotResults(uint256 _ballotId) external view returns(
+        uint256[] memory choicesWeighting,
+        uint256[] memory noOfChoicesInProposal,
+        address[] memory voters
+    )
+    {
+        if (_ballotId >= ballots.length 
+            || getCurrentBallotStage(_ballotId) != Stage.RESOLVED
+            || ballots[_ballotId].isCancelled
+        )
+            return (choicesWeighting, noOfChoicesInProposal, voters);
+        else {
+            address[] memory allowedVoters = getAllowedVotersByBallot(_ballotId);
+            Ballot storage ballot = ballots[_ballotId];
+            voters = new address[](ballot.totalVoters);
+            (uint256 i, uint256 j, uint256 k, uint256 count) = (0, 0, 0, 0);
+            // Filtering the actual voters address from the allowed voters address list
+            for (i = 0; i < allowedVoters.length; i++) {
+                if (ballot.voteDetails[allowedVoters[i]].voteOptions[0].length > 0) {
+                    voters[count] = allowedVoters[i];
+                    count++;
+                } 
+            }
+            count = 0;
+            // calculating the length of the choicesWeighting array it should be equal
+            // to the aggregation of ∑(proposalᵢ * noOfChoicesᵢ) where i = 0 ...totalProposal-1 
+            for (i = 0; i < ballot.totalProposals; i++) {
+                uint256 _noOfChoice = ballot.proposals[i].noOfChoices;
+                _noOfChoice = _noOfChoice == 0 ? 3 : _noOfChoice;
+                count = count.add(_noOfChoice);
+            }
+            choicesWeighting = new uint256[](count);
+            noOfChoicesInProposal = new uint256[](ballot.totalProposals);
+            count = 0;
+            for (i = 0; i < ballot.totalProposals; i++) {
+                uint256 _noOfChoices = ballot.proposals[i].noOfChoices;
+                _noOfChoices = _noOfChoices == 0 ? 3 : _noOfChoices;
+                noOfChoicesInProposal[i] = _noOfChoices;
+                uint256 nextProposalChoiceLen = count + _noOfChoices;
+                for (j = 0; j < ballot.totalVoters; j++) {
+                    uint256[] storage choiceWeight = ballot.voteDetails[voters[j]].voteOptions[i];
+                    uint256 l = 0;
+                    for (k = count; k < nextProposalChoiceLen; k++) {
+                        choicesWeighting[k] = choicesWeighting[k].add(choiceWeight[l]);
```

Comment:
I think this means that it is unlikely we would be able to retrieve the ballot results on-chain (assuming a large number of voters) - is it possible to keep some totals as people vote so that we could retrieve the results on-chain (e.g. so that a vote could be used to trigger some other on-chain action)?

Reply:
Yes, you're right it is not suitable to call this function on-chain. I don't think any easy and clean way to do it beside this - 
Rather keeping the weight of the choices per investor we will in parallel calculating the result (sum of weight for a choice of a proposal) of the vote and store it so when we will query the `getBallotResults()` function we will just return it. The downside of this mechanism is the increased gas cost of `revealVote()` function and It may change the functionality of many getters that are placed in the module.